### PR TITLE
aviator -diagram

### PR DESCRIPTION
**Pull request by rgarcia**
`aviator -diagram clever-api | dot -Tpng -o out.png`:

out

WIP, but would like any feedback before writing tests, etc.

cc @nathanleiby @dfuentes @alexandersoto

### CODE REVIEW

**Comment by dfuentes** [comment]
lgtm so far, I would probably put all the diagram code in its own file, since there doesn't seem to be any overlap with other commands.

**Comment by alexandersoto** [comment]
This is awesome! ensureags has two arrows to cleverenv, are there repeats being created? I'm not super familiar with go yet.

File: aviator/aviator.go [hunk 1]
```diff
@@ -132,6 +132,83 @@ func (a *Aviator) Launch(service string, defaults bool) error {
...
+			return err
+		}
+		for _, child := range children {
+			if err := crawl(child); err != nil {
+				return err
+			}
+		}
+		serviceCrawled[service.Name] = struct{}{}
+		if service.Name != s.Name {
+			services = append(services, service)
```
Comment (alexandersoto) [inline aviator/aviator.go]: Is this an append to a set or a list? If it's a list you might run into duplicates

File: aviator/aviator.go [hunk 2]
```diff
@@ -132,6 +132,83 @@ func (a *Aviator) Launch(service string, defaults bool) error {
...
+		}
+		return nil
+	}
+	return services, crawl(s)
+}
+
+// Children returns all the direct dependencies of a service.
+func (a Aviator) Children(s Service) ([]Service, error) {
+	services := []Service{}
```
Comment (nathanleiby) [inline aviator/aviator.go]: Likewise, could use a set here to simplify handling for dups

Comment (nathanleiby) [inline aviator/aviator.go]: https://github.com/fatih/set

File: aviator/aviator.go [hunk 3]
```diff
@@ -132,6 +132,83 @@ func (a *Aviator) Launch(service string, defaults bool) error {
...
+			} else {
+				services = append(services, requiredService)
+			}
+		}
+	}
+	return services, nil
+}
+
+func dotlabel(str string) string {
+	str = strings.Replace(str, "-", "", -1)
```
Comment (nathanleiby) [inline aviator/aviator.go]: style question-- is longer form preferred vs `return strings.Replace(str, "-", "", -1)`

**Comment by dfuentes** [comment]
I actually kind of like the dups, because it shows how strongly connected a dependency is in a way.  Perhaps we could use number of repeated references to a dependency as the line weight in the graph (or just put the weight as a label on the line).  Or we could dedup.

File: aviator/aviator.go [hunk 4]
```diff
@@ -132,6 +132,83 @@ func (a *Aviator) Launch(service string, defaults bool) error {
 	return a.environment.Services.Run()
 }
 
+// Dependencies returns all the services that a service depends on, in no particular order.
+func (a Aviator) Dependencies(s Service) ([]Service, error) {
+	services := []Service{}
+	serviceCrawled := map[string]struct{}{}
+	var crawl func(Service) error
+	crawl = func(service Service) error {
+		if _, alreadyCrawled := serviceCrawled[service.Name]; alreadyCrawled {
```
Comment (nathanleiby) [inline aviator/aviator.go]: Could you just use `services`, and get rid of `serviceCrawled` entirely? 

If using sets, check `services.Has(service)` here and on line 154

File: aviator/aviator.go (hunk 2 above)
Comment (rgarcia) [inline aviator/aviator.go]: a set of `Service` objects would be fairly complicated--I think you'd have to implement `set.Interface`: https://godoc.org/gopkg.in/fatih/set.v0#Interface. I think I prefer the simplicity of the slice

File: aviator/aviator.go (hunk 3 above)
Comment (rgarcia) [inline aviator/aviator.go]: this is an oversight--when I wrote this I originally thought I'd have to make more than one replacement: 87e1a60

File: aviator/aviator.go (hunk 1 above)
Comment (rgarcia) [inline aviator/aviator.go]: A list. Nice catch--this will cause duplicates. Will fix this when I write tests.

**Comment by rgarcia** [comment]
@dfuentes I'll move it to a separate file--my hope is that in the future we consolidate dependency crawling, since the code in this PR somewhat duplicates dependency crawling that happens elsewhere in the file (specifically here: https://github.com/Clever/aviator/blob/eb40291410e657d1f1c9e25093e9158763c67407/aviator/aviator.go#L184-L231). The direction I took with #48 was to consolidate crawling to just the `Dependencies` function, but it involves getting the dependencies in topological order, etc. which I think you expressed concerns about.

**Comment by rgarcia** [comment]
@dfuentes fixed dupe edge issue and added tests. New diagram output:

out

Assigning to you for review--thanks!

**Comment by dfuentes** [comment]
code looks good to me.  Sorry to harp on this, but could we name the files simply diagram and diagram_test?  aviator/aviator_diagram.go seems redundant.

**Comment by rgarcia** [comment]
@dfuentes 8125d3c

**Pull request merged**